pkg/proxy/healthcheck/healthcheck.go

```diff
@@ -20,108 +20,214 @@ import (
+}
+
+func (hcs *server) SyncServices(newServices map[types.NamespacedName]uint16) error {
+	hcs.lock.Lock()
+	defer hcs.lock.Unlock()
+
+	// Remove any that are not needed any more.
+	for nsn, svc := range hcs.services {
+		if port, found := newServices[nsn]; !found || port != svc.port {
+			glog.V(2).Infof("closing healthcheck %q on port %d", nsn.String(), svc.port)
+			if err := svc.listener.Close(); err != nil {
+				glog.Errorf("Close(%v): %v", svc.listener.Addr(), err)
+			}
+			delete(hcs.services, nsn)
+			delete(hcs.endpoints, nsn)
 		}
-		return
 	}
-	var entry *serviceEndpointsList
-	e, exists := h.serviceEndpointsMap.Get(req.serviceName.String())
-	if exists {
-		entry = e.(*serviceEndpointsList)
-		if entry.endpoints.Equal(*req.endpointUids) {
-			return
-		}
-		// Compute differences just for printing logs about additions and removals
-		deletedEndpoints := entry.endpoints.Difference(*req.endpointUids)
-		newEndpoints := req.endpointUids.Difference(*entry.endpoints)
-		for _, e := range newEndpoints.List() {
-			glog.V(4).Infof("Adding local endpoint %s to LB health check for service %s",
-				e, req.serviceName.String())
+
+	// Add any that are needed.
+	for nsn, port := range newServices {
+		if hcs.services[nsn] != nil {
+			glog.V(3).Infof("existing healthcheck %q on port %d", nsn.String(), port)
+			continue
 		}
-		for _, d := range deletedEndpoints.List() {
-			glog.V(4).Infof("Deleted endpoint %s from service %s LB health check (%d endpoints left)",
-				d, req.serviceName.String(), len(*entry.endpoints))
+
+		glog.V(2).Infof("opening healthcheck %q on port %d", nsn.String(), port)
+		svc := &hcInstance{port: port}
+		addr := fmt.Sprintf(":%d", port)
+		svc.server = hcs.httpFactory.New(addr, hcHandler{name: nsn, hc: hcs})
+		var err error
+		svc.listener, err = hcs.listener.Listen(addr)
+		if err != nil {
+			msg := fmt.Sprintf("node %s failed to start healthcheck %q on port %d: %v", hcs.hostname, nsn.String(), port, err)
+
+			if hcs.recorder != nil {
+				hcs.recorder.Eventf(
+					&clientv1.ObjectReference{
+						Kind:      "Service",
+						Namespace: nsn.Namespace,
+						Name:      nsn.Name,
+						UID:       types.UID(nsn.String()),
+					}, api.EventTypeWarning, "FailedToStartHealthcheck", msg)
+			}
+			glog.Error(msg)
+			continue
 		}
+		hcs.services[nsn] = svc
+
+		go func(nsn types.NamespacedName, svc *hcInstance) {
+			// Serve() will exit when the listener is closed.
+			glog.V(3).Infof("starting goroutine for healthcheck %q on port %d", nsn.String(), svc.port)
+			if err := svc.server.Serve(svc.listener); err != nil {
+				glog.V(3).Infof("healthcheck %q closed: %v", nsn.String(), err)
+				return
+			}
+			glog.V(3).Infof("healthcheck %q closed", nsn.String())
+		}(nsn, svc)
 	}
-	entry = &serviceEndpointsList{serviceName: req.serviceName, endpoints: req.endpointUids}
-	h.serviceEndpointsMap.Add(req.serviceName.String(), entry)
+	return nil
+}
+
+type hcInstance struct {
```

Comment:
naming is hard.  It's not a server, it's not a listener.  It's not actually doing checks, so it is not a checker.  beg for this to be good enough or fixed later?